### DISCUSSION TITLE
New Package: CSS Utility function

### DISCUSSION THREAD

**Discussion by schilchSICKAG** [synergy-design-system/synergy-design-system#278] [RFCs]
This RFC is based on the features discussed in #269. It provides an idea how to create a new package that will house the typographic styles, as well as others in the future.

## Naming

I suggest we use either `@synergy-design-system/css` OR `@synergy-design-system/utilities` as the new packages name. This is done as I would like to provide additional features in the future, too

---

## Package Contents

The package will initially only provide the typographic tokens discussed in #269. However, it may later on ship with various often needed utility functions and assets, for example:

- Default css bootstrapping for projects (e.g. a stylesheet which sets the fonts and correct sizes, as well as backgrounds and colors for everything once included).
- The fonts that are used in the design system. Currently, we do NOT ship the required fonts, which may lead to problems as every project will have to fetch them by itself.

---

## File system overview

- README.md
- CHANGELOG.md
- package.json
- src
  - bootstrap.css (pcss?)
  - typography.css (pcss?)
  - reset.css (potentially, not needed in first iteration!)
  - fonts/*.woff2 (All needed fonts)
  - fonts.css (pcss?)
- dist
  - bootstrap.css
  - full.css (see below in Final distribution files)

---

## full.css

The package should ship a file named `full.css` as its export. This file combines the reset (optional!), bootstrapping, typo and inlined fonts.

All other packages in `src` will also be transformed and provided in the module.

---

### Loading everything

```javascript
// @todo: Maybe we could also include the tokens in the css package?
import '@synergy-design-system/tokens/themes/light.css';

// Simple version. Bootstrapping, Typo and fonts just work
import '@synergy-design-system/css/full.css';
```

```javascript
// Must be loaded to use the tokens!
import '@synergy-design-system/tokens/themes/light.css';

// If we only want the typo, this is also possible
import '@synergy-design-system/css/typography.css';
```

---

## Releases

The package needs to be part of our automatic release cycle, making it mandatory to setup `semantic-release-monorepo` for the new package.

---

**Comment by kirchsuSICKAG** [comment]
I like your proposal. Here are some thoughts on the open things:

## Thoughts about package name
Will we only have css utilities or could there also be others? If yes I would go with the `@synergy-design-system/utilities` otherwise with `@synergy-design-system/css`

## Thoughts about css vs pcss
I really like nesting css styles because I think it is better readable in most cases. Therefore I would prefer pcss.
But if I got it right on MDN since last year all browsers support the `& nesting operator`.
But also for other features it could make sense and it is easier to start with pcss, than to start with css and afterwards went over to pcss.
Also for the reset.css we could use this plugin: https://github.com/csstools/postcss-normalize

## Thoughts about loading the theme directly in the full.css
I like the idea, so the users do not have to remember about importing it by themselves. 
But what if someone uses the full.css and wants to do something like this for switching the theme: 
![image](https://github.com/synergy-design-system/synergy-design-system/assets/115139104/e8455adb-1a98-499d-8257-f12f675a7d0a)

Would the color switch still work or would we override it always with the light theme, because we imported it explicitly?

**Reply by j-o-sh** [reply]
@kirchsuSICKAG I just had a discussion with @schilchSICKAG  about the `prefers-color-scheme` idea and we though it might be good to provide this as a separately named theme (something like _auto_ or _system_). 

The theme would basically include the 6 lines of code you proposed but would allow for app developers to actively opt in to automatic theme discovery.

Also you should look into browser engines whether you can go with the two `@media` queries... I can imagine some browsers don't set any value for the `prefers-color-scheme` variable and you would need to load a default for that. 🤔

`@synergy-design-system/tokens/dist/themes/auto.css`
```css
@import url('@synergy-design-system/tokens/dist/themes/light.css')
@media (prefers-color-schema: dark) {
  @import url('@synergy-design-system/tokens/dist/themes/dark.css')
}
```

But no, the cold switch wouldn't work like that because one css file would always be missing. 🤔

**Reply by schilchSICKAG** [reply]
I also think we should opt for pcss here. It would really be easy to create various output files like:

```bash
+ build/
 \- build.js # Main entry point for postcss
 \- auto.js # Custom transforms for the auto theme
+ src/
 \- mode.css
 \- forms.css
 \- typo.css
 \- reset.css
 \- fonts.css
 \- baseline.css
+ dist/
 \- full.css # includes themes (from tokens), reset, fonts, baseline
 \- auto.css # includes themes (via custom match-media), reset, fonts, baseline
+ fonts/
 \- opensans.woff2 # and all the other fallbacks
```

What do you think? For the naming perspective, I would like to call it `@synergy-design-system/css`. This is because I think if we **would** ship special functionality that is required in our components code, we should add it either

a) directly add it to `@synergy-design-system/components` or
b) ship this (theoretic) new package together with `@synergy-design-system/components` anyway.

**Reply by schilchSICKAG** [reply]
> Would the color switch still work or would we override it always with the light theme, because we imported it explicitly?

I think @j-o-sh is right when stating that we should do this with a custom import that we provide (the mentioned `auto.css`, until we find a more descriptive name).

**Reply by kirchsuSICKAG** [reply]
> What do you think? For the naming perspective, I would like to call it @synergy-design-system/css. This is because I think if we would ship special functionality that is required in our components code, we should add it either

You're right, I'm fine with the naming for the package `@synergy-design-system/css`.
Also if we add for example the css for tables at some point it also fits as it will be css only.

**Comment by schilchSICKAG** [comment]
I think we will also have to create proper documentation for those features, e.g. which package to use when because you can either:

a) use `@synergy-design-system/tokens` as we are documenting it currently or
b) use `@synergy-design-system/tokens` and `@synergy-design-system/css` side by side and also
c) use `@synergy-design-system/css` exclusively.

Each has its own pros and cons. Examples:

| Package | Includes | When to use
|:--------|:---------|:-------------
| `tokens`  | Tokens in CSS (always needed), JS and SCSS | When you want to build your own theme (e.g. via JS exports) or don´t care about our typographic and bootstrapping styles
| `css` | Only css helper utilities, including the vars from `tokens` | Use this if you do not need the SCSS or JS exports for low level tokens at all.
| `tokens`&`css` | Everything | When you want to apply the basic recommended styling and also need the css vars inline (e.g. via Javascript `Element.style`)

**Comment by schilchSICKAG** [comment]
Please have a look at https://github.com/synergy-design-system/synergy-design-system/tree/feat/269-add-css-utility-package for a first draft. Would love to hear your opinions on this.

**Reply by klaarseSICKAG** [reply]
Instead of calling the package CSS I'd rather use a name which describes more the semantic instead of the used technology. Otherwise wie have tokens -> CSS and CSS -> CSS, which I find a bit confusing, and the user also don't know what to expect from theses packages. I think of smth. like BASE_STYLING or similar. Would also have the advantage that we can add smth. different than CSS later on...

**Reply by schilchSICKAG** [reply]
I think you are right with this. Also, the `css` package will also export more than exports later on (e.g. fonts), so the name still does not fit quite.

Just to get some ideas together:

`@synergy-design-system/layout`
`@synergy-design-system/style`
`@synergy-design-system/styles`
`@synergy-design-system/base-styles`
`@synergy-design-system/bootstrap`
`@synergy-design-system/foundations`

**Reply by klaarseSICKAG** [reply]
- layout: doesn't fit to typography!?
- foundations: are the tokens not also foundations? what would be the difference in meaning?
- style/styles: might be too generic (very similar to css)
- base-styles: also a bit generic but somehow I like this best
- bootstrap: collides a bit with the popular css framework and might suggest smth unexpected. not sure though...
- other suggestions: primitives, core-styles, common-styles, root-styles

**Reply by reichst1SICKAG** [reply]
I would opt for `base-styles` because this is what we provide here.

CSS is also not really wrong. In the end the user is including *.css into his files.

**Reply by schilchSICKAG** [reply]
I suppose we should discuss this tomorrow in our team meeting face to face. I am also not quite sure about the naming anymore :).

**Comment by reichst1SICKAG** [comment]
Do we also want to include a normalize.css or a reset.css for a baseline setting?
I think that normalize.css is way better than a reset.css but ...

.. [do we really need that?](https://jaydenseric.com/blog/forget-normalize-or-resets-lay-your-own-css-foundation)

Also **HTML5Boilerplate** has some [base settings](https://github.com/h5bp/main.css/tree/main/src) which compiles to https://github.com/h5bp/html5-boilerplate/blob/main/dist/css/style.css looking nice and usable.

**Comment by schilchSICKAG** [comment]
This will now be handled via the steps as defined in #365.
